cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkUtils.java

```diff
@@ -155,6 +188,158 @@ public static URLClassLoader createSparkFrameworkClassLoader(ClassLoader parentC
+    }
+  }
+
+  /**
+   * Rewrites the bytecode of the org.apache.spark.deploy.yarn.Client class to fix the bug in SPARK-13441
+   *
+   * @param input {@link InputStream} for reading the original bytecode of the Client class.
+   * @return The rewritten bytecode
+   * @throws IOException if failed to read from the given input
+   */
+  private static byte[] rewriteSparkYarnClient(InputStream input) throws IOException {
+    ClassReader cr = new ClassReader(input);
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+    cr.accept(new ClassVisitor(Opcodes.ASM5, cw) {
+      @Override
+      public MethodVisitor visitMethod(final int access, final String name,
+                                       final String desc, String signature, String[] exceptions) {
+        MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+
+        // Only rewrite the createConfArchive method
+        if (!"createConfArchive".equals(name)) {
+          return mv;
+        }
+
+        // Generate the method body to just return the hardcode LOCALIZED_CONF_DIR_ZIP file name
+        if (Type.getReturnType(desc).equals(Type.getType(File.class))) {
+          // Spark 1.5+ return type is File
+          // This is what gets generated
+          // return new File(LOCALIZED_CONF_DIR_ZIP);
+          GeneratorAdapter mg = new GeneratorAdapter(mv, access, name, desc);
+          mg.newInstance(Type.getType(File.class));
+          mg.dup();
+          mg.visitLdcInsn(LOCALIZED_CONF_DIR_ZIP);
+          mg.invokeConstructor(Type.getType(File.class), Methods.getMethod(void.class, "<init>", String.class));
+          mg.returnValue();
+          mg.endMethod();
+          return null;
+        } else if (Type.getReturnType(desc).equals(Type.getType(Option.class))) {
+          // Spark 1.4 return type is Option<File>
+          // This is what gets generated
+          // return Option.apply(new File(LOCALIZED_CONF_DIR_ZIP);
+          GeneratorAdapter mg = new GeneratorAdapter(mv, access, name, desc);
+          mg.newInstance(Type.getType(File.class));
+          mg.dup();
+          mg.visitLdcInsn(LOCALIZED_CONF_DIR_ZIP);
+          mg.invokeConstructor(Type.getType(File.class), Methods.getMethod(void.class, "<init>", String.class));
+          mg.invokeStatic(Type.getType(Option.class), Methods.getMethod(Object.class, "apply", Object.class));
+          mg.checkCast(Type.getType(Option.class));
+          mg.returnValue();
+          mg.endMethod();
+          return null;
+        }
+        return mv;
+      }
+    }, 0);
+    return cw.toByteArray();
+  }
+
+  /**
+   * Returns the local temporary directory as specified by the configuration.
+   */
+  private static File getTempDir(CConfiguration cConf) {
+    File tempDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    tempDir.mkdirs();
+    return tempDir;
```

Comment:
Also, in particular to this change, we only create one temp file per master process life time.